File: notes-n-resources/Data-Structures-N-Algo/_DS-n-Algos/__MY_OPRIGINAL_DS/_Extra-Practice/06_breadth-first_search/ts/01_breadth-first_search.ts
interface Graph<T> {
    [key: string]: T[]
};

function personIsSeller(name: string): boolean {
    return name.endsWith('m');
}

const graph: Graph<string> = {};
graph.you = ['alice', 'bob', 'claire'];
graph.bob = ['anuj', 'peggy'];
graph.alice = ['peggy'];
graph.claire = ['thom', 'jonny'];
graph.anuj = [];
graph.peggy = [];
graph.thom = [];
graph.jonny = [];

function search(name: string): boolean {
    let searchQueue = graph[name];
    // This array is how you keep track of which people you've searched before.
    const searched = [];

    while(searchQueue.length > 0) {
        const person = searchQueue.shift();
        // Only search this person if you haven't already searched them
        if (!searched.includes(person)) {
            if (personIsSeller(person)) {
                console.log(`${person} is a mango seller!`);
                return true;
            } else {
                searchQueue = searchQueue.concat(graph[person]);
                // Marks this person as searched
                searched.push(person);
            }
        }
    }

    return false;
};

search('you');

File: notes-n-resources/Data-Structures-N-Algo/_DS-n-Algos/__MY_OPRIGINAL_DS/_Extra-Practice/06_breadth-first_search/ruby/01_breadth-first_search.rb
def person_is_seller?(name)
  name[-1] == "m"
end

class Graph < Hash
  def search(name)
    search_queue = []
    search_queue += self[name]
    # The "searched" Hash is how you keep track of which people you've searched before. We use a hash because hash lookups are fast!
    searched = {}

    until search_queue.empty?
      person = search_queue.shift
      # Only search this person if you haven't already searched them.
      next if searched[person]
      return person if yield person
      search_queue += self[person]
      # Marks this person as searched
      searched[person] = true
    end

    false
  end
end


# %w(string1 string2 ...) is a shorter way to define arrays of strings
graph = Graph.new
graph["you"] = %w(alice bob claire)
graph["bob"] = %w(anuj peggy)
graph["alice"] = %w(peggy)
graph["claire"] = %w(thom jonny)
graph["anuj"] = []
graph["peggy"] = []
graph["thom"] = []
graph["jonny"] = []

# we begin the search from vertex "you" and pass the match criterion in the block, then we tap the search result
graph.search("you"){|person| person_is_seller?(person)}.tap{|person| puts "#{person} is a mango seller!" if person}
